staging/src/k8s.io/apiserver/pkg/server/openapi/openapi_aggregator.go

```diff
@@ -0,0 +1,462 @@
+		if schema.Items.Schema != nil {
+			s.walkSchema(schema.Items.Schema)
+		}
+		for _, v := range schema.Items.Schemas {
+			s.walkSchema(&v)
+		}
+	}
+}
+
+func (s *walkAllRefs) walkParams(params []spec.Parameter) {
+	if params == nil {
+		return
+	}
+	for _, param := range params {
+		param.Ref = s.walkRef(param.Ref)
+		s.walkSchema(param.Schema)
+		if param.Items != nil {
+			param.Items.Ref = s.walkRef(param.Items.Ref)
+		}
+	}
+}
+
+func (s *walkAllRefs) walkResponse(resp *spec.Response) {
+	if resp == nil {
+		return
+	}
+	resp.Ref = s.walkRef(resp.Ref)
+	s.walkSchema(resp.Schema)
+}
+
+func (s *walkAllRefs) walkOperation(op *spec.Operation) {
+	if op == nil {
+		return
+	}
+	s.walkParams(op.Parameters)
+	if op.Responses == nil {
+		return
+	}
+	s.walkResponse(op.Responses.Default)
+	for _, r := range op.Responses.StatusCodeResponses {
+		s.walkResponse(&r)
+	}
+}
+
+func (s *walkAllRefs) Start() {
+	for _, pathItem := range s.root.Paths.Paths {
+		s.walkParams(pathItem.Parameters)
+		s.walkOperation(pathItem.Delete)
+		s.walkOperation(pathItem.Get)
+		s.walkOperation(pathItem.Head)
+		s.walkOperation(pathItem.Options)
+		s.walkOperation(pathItem.Patch)
+		s.walkOperation(pathItem.Post)
+		s.walkOperation(pathItem.Put)
+	}
+}
+
+// FilterSpecByPaths remove unnecessary paths and unused definitions.
+func FilterSpecByPaths(sp *spec.Swagger, keepPathPrefixes []string) {
+	// First remove unwanted paths
+	prefixes := trie.New(keepPathPrefixes)
+	orgPaths := sp.Paths
+	if orgPaths == nil {
+		return
+	}
+	sp.Paths = &spec.Paths{
+		VendorExtensible: orgPaths.VendorExtensible,
+		Paths:            map[string]spec.PathItem{},
+	}
+	for path, pathItem := range orgPaths.Paths {
+		if !prefixes.HasPrefix(path) {
+			continue
+		}
+		sp.Paths.Paths[path] = pathItem
+	}
+
+	// Walk all references to find all definition references.
+	usedDefinitions := map[string]bool{}
+	newWalkAllRef(func(ref spec.Ref) spec.Ref {
+		if ref.String() != "" {
+			refStr := ref.String()
+			if strings.HasPrefix(refStr, DEFINITION_PREFIX) {
+				usedDefinitions[refStr[len(DEFINITION_PREFIX):]] = true
+			}
+		}
+		return ref
+	}, sp).Start()
+
+	// Remove unused definitions
+	orgDefinitions := sp.Definitions
+	sp.Definitions = spec.Definitions{}
+	for k, v := range orgDefinitions {
+		if usedDefinitions[k] {
+			sp.Definitions[k] = v
+		}
+	}
+}
+
+func equalSchemaMap(s1, s2 map[string]spec.Schema) bool {
+	if len(s1) != len(s2) {
+		return false
+	}
+	for k, v := range s1 {
+		v2, found := s2[k]
+		if !found {
+			return false
+		}
+		if !EqualSchema(&v, &v2) {
+			return false
+		}
+	}
+	return true
+}
+
+func equalSchemaArray(s1, s2 []spec.Schema) bool {
+	if s1 == nil {
+		return s2 == nil
+	}
+	if len(s1) != len(s2) {
+		return false
+	}
+	for _, v1 := range s1 {
+		found := false
+		for _, v2 := range s2 {
+			if EqualSchema(&v1, &v2) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	for _, v2 := range s2 {
```

Comment:
If we have duplicate values, that won't work. e.g. ["a" "a" "a"] and ["a" "b" "c"] are equal if we just take value from first list and find them in the second one. There could be other ways to solve this problem (removing items, etc.) but I think going both way works too.